src/storage/txn/actions/check_txn_status.rs

```diff
@@ -49,20 +135,23 @@ pub fn check_txn_status_lock_exists(
     }
 
     let is_pessimistic_txn = !lock.for_update_ts.is_zero();
-    if lock.ts.physical() + lock.ttl < current_ts.physical() {
-        // If the lock is expired, clean it up.
-        // If the resolving and primary key lock are both pessimistic locks, just unlock
-        // the primary pessimistic lock and do not write rollback records.
-        return if resolving_pessimistic_lock && lock.lock_type == LockType::Pessimistic {
-            let released = txn.unlock_key(primary_key, is_pessimistic_txn, TimeStamp::zero());
-            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.pessimistic_rollback.inc();
-            Ok((TxnStatus::PessimisticRollBack, released))
-        } else {
-            let released =
-                rollback_lock(txn, reader, primary_key, &lock, is_pessimistic_txn, true)?;
-            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.rollback.inc();
-            Ok((TxnStatus::TtlExpire, released))
-        };
+    if lock.is_pessimistic_lock() {
+        let (status, released) = check_txn_status_from_pessimistic_primary_lock(
```

Comment:
Good question, In my opinion, for the newer version, the kv client should always set it to true as indicated in https://github.com/tikv/tikv/blob/master/src/storage/txn/commands/check_txn_status.rs#L57. For older versions, resolving the primary-key switching issue was not possible.

Another potential solution is to never place a `rollback` record on the **pessimistic lock** primary key. This would prevent breaking transaction status. However, if during prewrite phase, the prewrite request on the primary key is lost and remains a pessimistic lock while secondary locks are resolved as prewrite ones, it could neither be rolled back nor committed since we cannot determine its transaction status.

/cc @MyonKeminta

Reply:
Actually, my original idea is to check write cf to see if there is any record of this transaction if check_txn_status finds the primary has a pessimistic lock, *and* it's marked to be locked with conflict.

Reply:
Yes, it's much more complicated than before because we have to make sure if the current `primary key` and `lock on the key` are valid, the transaction status should be determined only if the information are valid.